Newsletter Settings: Add Newsletter Categories settings card with a toggle

andregardi opened the pull request:
At the Newsletter setting page, there is the **Newsletter categories** section.

As part of the effort to untangle calypso from wp-admin, we should recreate this section on Jetpack Newsletter setting page:
https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/_inc/client/newsletter/index.jsx

This PR is a work in progress under the `enable-newsletter-categories` feature flag, that allows us to toggle the categories on/off.
![image](https://github.com/Automattic/jetpack/assets/33497086/f6e3abc6-459b-4107-9bc2-b2f71ecdf986)

This also includes a raw text input for editing which categories the users checked. It is not meant for the final user, we use it only to test the connection of the interface with backend.

## Does this pull request change what data or activity we track or use?
No

## Testing instructions:
To enable the feature flag navigate with **enable-newsletter-categories** query param:
`wp-admin/admin.php?enable-newsletter-categories=true&page=jetpack#/newsletter`

**Testing the toggle:**
- Toggle the setting on jetpack side.
- Check that the same option was toggle on wordpress.com for that site.

**Testing the raw input for categories check:**
![image](https://github.com/Automattic/jetpack/assets/33497086/70fc5af1-7608-4964-9892-cbc1a5441d73)

- On wordpress.com side, check about 3 categories and save the setting.
- On jetpack side, copy the array from `Categories state` field to `New value`, remove one of the ids from the array.
- Click on `Parse array to state `. It should save the new checked list.
- Refresh the wordpress.com page. It should reflect the new state in which one of the categories were unchecked.

Commit:
Api connection to newsletter categories.

Commit:
changelog

Commit:
Put feature under enable-newsletter-categories flag.

Commit:
Add raw categories JSON to UI.

phcp commented on projects/plugins/jetpack/_inc/client/newsletter/newsletter-categories.jsx:
```diff
@@ -0,0 +1,86 @@
+import TextInput from '../components/text-input';
+import Textarea from '../components/textarea';
+
+const mapCategoriesIds = categorie => {
```
Typo on `category`, would be nice to fix before merging.

andregardi replied:
Nice catch! It is fixed now.

phcp reviewed (dismissed):
Nice work! The code looks good and on my tests the save and sync worked as expected. There is just one typo that would be nice to fix before merging.

Commit:
Fix typo.

mashikag commented on projects/plugins/jetpack/_inc/client/newsletter/newsletter-categories.jsx:
```diff
@@ -0,0 +1,86 @@
+
+	const [ newCategories, setNewCategories ] = useState( '' );
+
+	const handleEnagleNewsletterCategoriesToggleChange = useCallback( () => {
```
```suggestion
	const handleEnableNewsletterCategoriesToggleChange = useCallback( () => {
```

mashikag commented on projects/plugins/jetpack/_inc/client/newsletter/newsletter-categories.jsx:
```diff
@@ -0,0 +1,86 @@
+			<SettingsGroup hasChild disableInOfflineMode disableInSiteConnectionMode>
+				<ToggleControl
+					checked={ isNewsletterCategoriesEnabled }
+					onChange={ handleEnagleNewsletterCategoriesToggleChange }
```
```suggestion
					onChange={ handleEnableNewsletterCategoriesToggleChange }
```

mashikag commented on projects/plugins/jetpack/_inc/lib/class.core-rest-api-endpoints.php:
```diff
@@ -2643,6 +2643,20 @@ public static function get_updateable_data_list( $selector = '' ) {
+			'wpcom_newsletter_categories_enabled'  => array(
+				'description'       => esc_html__( 'Whether the newsletter categories are enabled or not', 'jetpack' ),
+				'type'              => 'boolean',
+				'default'           => 1,
```
```suggestion
				'default'           => 0,
```

mashikag commented:
![localhost_wp-admin_admin php_page=jetpack enable-newsletter-categories=true](https://github.com/Automattic/jetpack/assets/2019970/6de4fb63-4147-44cc-852b-22e84fc9e342)

The setting remains enabled in the offline mode. It is expected for it to be disabled.

mashikag commented on projects/plugins/jetpack/_inc/client/newsletter/newsletter-categories.jsx:
```diff
@@ -0,0 +1,86 @@
+
+	return (
+		<SettingsCard { ...props } module="subscriptions">
+			<SettingsGroup hasChild disableInOfflineMode disableInSiteConnectionMode>
```
Why `disableInSiteConnectionMode`? 🤔

mashikag replied:
Looks like we do need it. Apparently it is possible to connect a self hosted site to Jetpack without setting up WPCOM account (p1677772292841239-slack-CBG1CP4EN). When that happens we do not say that the site is WPCOM connected, but instead we call it site connected.

mashikag committed:
Disable the toggle when no connection or site only connection.

mashikag commented:
I wondered what is the desired behaviour with when to save the "Enable newsletter categories" toggle. Should we save it using the Save Changes button in the settings card, or should we auto save it on each toggle action (as per the current implementation). 🤔  

@crisbusquets , what is your take on this? It is getting a bit confusing like the Calypso settings pages.

Current implementation (WIP):
![Screen Capture on 2024-02-20 at 11-30-56](https://github.com/Automattic/jetpack/assets/2019970/b944f782-457f-4df5-9466-7d70bb14627c)

crisbusquets commented:
> I wondered what is the desired behaviour with when to save the "Enable newsletter categories" toggle. Should we save it using the Save Changes button in the settings card, or should we auto save it on each toggle action (as per the current implementation). 🤔
> 
> @crisbusquets , what is your take on this? It is getting a bit confusing like the Calypso settings pages.
> 

Replied tho this here: p1HpG7-r4W-p2#comment-69704

mashikag committed:
Pass module to SettingsGroup component; Introduce SUBSCRIPTIONS_MODULE_NAME constant;

mashikag committed:
Update projects/plugins/jetpack/_inc/client/newsletter/newsletter-categories.jsx

Fix typo

mashikag committed:
Update projects/plugins/jetpack/_inc/client/newsletter/newsletter-categories.jsx

Fix typo

mashikag committed:
Update projects/plugins/jetpack/_inc/lib/class.core-rest-api-endpoints.php

Change the default value of `wpcom_newsletter_categories_enabled` to `0` (`false`).

mashikag committed:
Hide save buttonfrom Newsletter categories settings card; Set the header of the Newsletter categories settings card;

mashikag committed:
Merge branch 'update/newsletter-categories-enabled-toggle' of github.com:Automattic/jetpack into update/newsletter-categories-enabled-toggle

mashikag committed:
Use SUBSCRIPTIONS_MODULE_NAME const where appropriate

mashikag reviewed (approved):
I am happy with the current state of the code. I tested it locally and with jurassic tube, most of the changes are behind the feature flag so it is safe to merge it. I approve. 👍
